### PR TITLE
add documentation for eb --extended-dry-run

### PR DESCRIPTION
**Pull request by boegel**
see https://github.com/hpcugent/easybuild-framework/pull/1388

preview available at http://boegel-eb.readthedocs.org/en/extended_dry_run/Extended_dry_run.html

should be fleshed out a bit more, including:

* [x] mentioning `--disable-extended-dry-run-ignore-errors`
* [x] more details about simulating of `module load` of missing modules (esp. the toolchain module)
* [x] guidelines for making easyblocks fully compatible with `--extended-dry-run`

### CODE REVIEW

File: docs/Extended_dry_run.rst [hunk 1]
```diff
@@ -0,0 +1,731 @@
...
+For example, expressions that are conditional on the presence of certain files or directories in the build directory
+will always be false, since the build directory is never actually populated.
+
+.. _extended_dry_run_notes_ignored_errors:
+
+Errors are ignored (by default) during dry run
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Any errors that occur are ignored, and are reported with a clear warning message.
+This is done because it is not unlikely that these errors occur because of the dry run mechanism.
```
Comment (rjeschmi) [inline docs/Extended_dry_run.rst]: double negative is harder to read. Perhaps change to "likely" or "possible" or "probable"

File: docs/Extended_dry_run.rst [hunk 2]
```diff
@@ -0,0 +1,731 @@
+.. _extended_dry_run:
+
+Extended dry run
+================
+
+Using ``--extended-dry-run`` or ``-x`` (supported since EasyBuild v2.4.0), a detailed overview of the build and install
```
Comment (rjeschmi) [inline docs/Extended_dry_run.rst]: Do we link to the changelog here?

File: docs/Extended_dry_run.rst [hunk 3]
```diff
@@ -0,0 +1,731 @@
...
+.. _extended_dry_run_overview_run_cmd:
+
+Shell commands are not executed
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Any shell commands that are executed via the ``run_cmd`` and ``run_cmd_qa`` functions that are provided by the
+EasyBuild framework via the ``easybuild.tools.run`` are *not* executed, only reported
+(see also :ref:`extended_dry_run_guidelines_easyblocks_framework_functions_run_cmd`).
+
+This typically includes the commands that are defined in the easyblockto be run as a part of the
```
Comment (rjeschmi) [inline docs/Extended_dry_run.rst]: typo easyblockto

File: docs/Extended_dry_run.rst [hunk 4]
```diff
@@ -0,0 +1,731 @@
...
+
+    file written: /tmp/eb-ksVC07/tmp.conf
+
+.. _extended_dry_run_guidelines_easyblocks_framework_functions_runtime_patching:
+
+Runtime patching of files: ``apply_regex_substitutions``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+To make runtime patching of files in easyblocks easier, and to do it with taking the possibility of being in dry run
+module into account, the EasyBuild framework provides the ``apply_regex_substitutions`` function (available from the
```
Comment (rjeschmi) [inline docs/Extended_dry_run.rst]: This should link to documentation about this feature.

Is this valid in an easyconfig?

File: docs/Extended_dry_run.rst [hunk 5]
```diff
@@ -0,0 +1,731 @@
...
+
+For this, the ``run_cmd`` function support another named argument, i.e. ``forced``.
+When set to ``True``, the specified command will always be executed, even when in dry run mode.
+
+This is mainly intended for use in the EasyBuild framework itself, where commands that verify certain things must
+be executed, but it can also be useful for easyblocks (if used correctly).
+
+For example::
+
+    out, exit_code = run_cmd("type module", simple=False, forced=True)
```
Comment (rjeschmi) [inline docs/Extended_dry_run.rst]: not necessarily a documentation issue, but I don't like the parameter being called "forced." Who knows if there will be a future even more forced option.

Maybe it should be "run_in_dry" or "forced_wet" :)

**Comment by rjeschmi** [comment]
looks fine, some typos.

I think the apply_regex_substitutions is an important feature to highlight, since it will probably become our "best practice" rather than one line patch files...

File: docs/Extended_dry_run.rst (hunk 1 above)
Comment (boegel) [inline docs/Extended_dry_run.rst]: ok, thanks

File: docs/Extended_dry_run.rst (hunk 2 above)
Comment (boegel) [inline docs/Extended_dry_run.rst]: You mean release notes?

I can, but they still need to be fleshed out for the upcoming v2.4.0.

I can get them started and only include an entry for https://github.com/hpcugent/easybuild-framework/pull/1388 though, so I will, and include a reference here.

File: docs/Extended_dry_run.rst (hunk 4 above)
Comment (boegel) [inline docs/Extended_dry_run.rst]: No, not yet, but at some point, we should look into adding support for listing regex entries in `patches`, see also https://github.com/hpcugent/easybuild-framework/issues/1400

File: docs/Extended_dry_run.rst (hunk 5 above)
Comment (boegel) [inline docs/Extended_dry_run.rst]: Hmm... I see your point, but I don't like your alternatives either. ;-)

How about naming it `dry_run_safe`? Or `also_in_dry_run_mode`?

Comment (boegel) [inline docs/Extended_dry_run.rst]: I went with `force_in_dry_run`, changed in docs and framework PR. Thanks!

**Comment by boegel** [comment]
Good to go now https://github.com/hpcugent/easybuild-framework/pull/1388 is merged in.

Thanks for the review @rjeschmi

**Pull request merged**